### PR TITLE
Make CMake work on Windows

### PR DESCRIPTION
**Pull request by moxian**
#### Summary
SUMMARY: Build "Make CMake generators work on Windows (MSYS and VS)"

#### Purpose of change
I failed to build the game with cmake from master - neither in MSYS environment, nor windows command prompt with Visual Studio generators worked (for different reasons).
This PR fixes that and makes the game at least buildable via cmake for those systems.

#### Describe the solution
Tweaked to CMakeLists.txt, fixed otherwise harmless code bugs.
This was a multi-step process of discovering new bugs when the old ones got fixed. See individual commit messages for detailed descriptions.

#### Describe alternatives you've considered
Leave things be, and use pre-existing Make / *.vcxproj files. 
While those solutions (kinda) work, I personally prefer CMake (due to its ability to build files out of tree, primarily), and having options open won't hurt, right?

#### Additional context
None.

### CODE REVIEW

**Comment by moxian** [comment]
Fixed summary line.

File: tests/rewrite_vsnprintf_test.cpp [hunk 1]
```diff
@@ -33,3 +36,5 @@ TEST_CASE( "Test vsnprintf_rewrite" )
            "<color_%s>powered welder</color> (and <color_%s>welding goggles</color>) or "
            "<color_%s>duct tape</color>, and level <color_%s>%d</color> skill in mechanics.%s%s" );
 }
+
+#endif
```
Comment (ZhilkinSerg) [inline tests/rewrite_vsnprintf_test.cpp]:
```suggestion
#endif

```

**Comment by ZhilkinSerg** [comment]
Jenkins rebuild.

**Comment by moxian** [comment]
The force-push above was removing WIN32 option on the binaries - I added it somewhat hastily, and didn't test it well enough, so it fails to build for some configurations in the state I've pushed it originally.

Rest of the commit chain is the same.

**Comment by ZhilkinSerg** [comment]
Can you please update [docs](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/COMPILING-CMAKE.md) too?

**Comment by moxian** [comment]
Updated, thanks for bringing this up.

While doing so I realized that not currently possible to build with MSBuild with either BACKTRACE=ON, LUA=ON or LOCALIZE=ON . It's possible *in principle* of course, and I had a super hacky solutions for all three of those in a local dirty repo, but they are all a bit more involved, and, again, I'd like to handle them separately.

**Comment by ZhilkinSerg** [comment]
Merged to `0.D` (and `master` with with #28097).

**Pull request closed without merging**